src/main/java/gregtech/api/recipes/ingredients/GTRecipeItemInput.java

```diff
@@ -0,0 +1,163 @@
+    }
+
+    public static IGTRecipeInput getOrCreate(ItemStack ri) {
+        return getFromCache(new GTRecipeItemInput(ri));
+    }
+
+    @Override
+    public int getAmount() {
+        return this.amount;
+    }
+
+    @Override
+    public GTRecipeItemInput setNonConsumable() {
+        GTRecipeItemInput ret = new GTRecipeItemInput(this.inputStacks, this.amount);
+        ret.isConsumable = false;
+        if (GTIngredientCache.NON_CONSUMABLE_INSTANCES.get(ret) == null) {
+            GTIngredientCache.NON_CONSUMABLE_INSTANCES.put(ret, new WeakReference<>(ret));
+        } else {
+            ret = GTIngredientCache.NON_CONSUMABLE_INSTANCES.get(ret).get();
+        }
+        return ret;
+    }
+
+    @Override
+    public IGTRecipeInput setNBTMatchingCondition(NBTMatcher nbtMatcher, NBTcondition nbtCondition) {
+        this.nbtMatcher = nbtMatcher;
+        this.nbtCondition = nbtCondition;
+        return this;
+    }
+
+    @Override
+    public boolean hasNBTMatchingCondition() {
+        return this.nbtMatcher != null;
+    }
+
+    @Override
+    public NBTMatcher getNBTMatcher() {
+        return nbtMatcher;
+    }
+
+    @Override
+    public NBTcondition getNBTMatchingCondition() {
+        return nbtCondition;
+    }
+
+    @Override
+    public boolean isNonConsumable() {
+        return !isConsumable;
+    }
+
+    @Override
+    public ItemStack getInputStack() {
+        return this.inputStacks;
+    }
+
+    @Override
+    public FluidStack getInputFluidStack() {
+        return null;
+    }
+
+    @Override
+    public boolean isOreDict() {
+        return false;
+    }
+
+    @Override
+    public boolean acceptsStack(ItemStack input) {
+        if (input == null) {
+            return false;
+        }
+        ItemStack stacks = this.inputStacks;
+        if (stacks.getItem() == input.getItem()) {
+            int i = stacks.getMetadata();
+            if (i == 32767 || i == input.getMetadata()) {
+                return (nbtMatcher == null ? ItemStack.areItemStackTagsEqual(stacks, input) : nbtMatcher.evaluate(stacks, nbtCondition));
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public int hashCode() {
+        ItemStack stack = getInputStack();
+        if (nbtMatcher == null) {
+            return Objects.hash(stack.getItem(), stack.getMetadata(), stack.getTagCompound());
+        }
+        return Objects.hash(stack.getItem(), stack.getMetadata(), 0);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (this == other) return true;
+        if (!(other instanceof GTRecipeItemInput)) return false;
+        GTRecipeItemInput otherIngredient = (GTRecipeItemInput) other;
+
+        if (this.amount != otherIngredient.amount) return false;
+        if (this.isConsumable != otherIngredient.isConsumable) return false;
+
+
+        if (!this.getInputStack().isItemEqual(otherIngredient.getInputStack())) return false;
```

Comment:
Should this check take into account if `this.getInputStack` has wildcard metadata? `isItemEqual` will only return true if `otherIngredient.getInputStack` also has wildcard metadata, but would we want this check to also be true if only the items are equal and `this.getInputStack` has Wildcard metadata?

Reply:
i think we dont actually care about the wildcard here since the oredict ingredient has a separate object